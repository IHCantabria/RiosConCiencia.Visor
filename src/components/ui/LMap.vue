<template>
  <div class="l-map">
    <div id="mapid" class="l-map__view" />
  </div>
</template>
<script>
import L from "leaflet";
import "leaflet-draw/dist/leaflet.draw.js";
import "leaflet-easybutton/src/easy-button.js";
import "leaflet-extra-markers/dist/js/leaflet.extra-markers.min.js";
import {
  iconFix,
  createCustomIcon,
  panZoomMarker
} from "@/utils/leaflet-utils.js";
iconFix();
export default {
  props: {
    cCenter: {
      type: Object,
      required: true
    },
    cBaseLayers: {
      type: Array,
      required: false,
      default: () => {
        return [
          {
            url:
              "https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}",
            configParameters: {
              id: "defaultMapID",
              attribution:
                '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
            }
          }
        ];
      }
    },
    cMapDefault: {
      type: Object,
      required: false,
      default: () => {
        return {
          center: [0, 0],
          zoom: 3,
          layers: ["defaultMapID"]
        };
      }
    },
    cDrawOptions: {
      type: Object,
      required: false,
      default: () => {
        return null;
      }
    },
    cSwitchLayerBaseMaps: {
      type: Array,
      required: false,
      default: () => {
        return [];
      }
    },
    cMarkers: {
      type: Array,
      required: false,
      default: () => {
        return [];
      }
    },
    goToCoords: {
      type: Object,
      required: false,
      default: () => {
        return null;
      }
    }
  },
  data() {
    return {
      myMap: null,
      layersMap: [],
      layerMarkers: [],
      editableLayers: null
    };
  },
  watch: {
    cMarkers: {
      handler() {
        this.updateLayerMarker();
      },
      deep: true
    },
    goToCoords: {
      handler(newValue) {
        newValue ? this.goTo(newValue) : "";
      },
      deep: true
    }
  },
  mounted() {
    this.$nextTick(() => {
      this.initMap();
    });
  },
  methods: {
    async initMap() {
      this.loadMap();
    },
    loadMap() {
      this.loadTileLayers();
      const layersToSetMap = this.layersMap.filter(layer =>
        this.cMapDefault.layers.includes(layer.options.id)
      );
      this.setMap(layersToSetMap);
      this.addGlobalSearchControl();
      this.cDrawOptions ? this.addDrawControls() : "";
      this.setupSwitcherLayer();
      this.setMarker();
      this.setupListeners();
    },
    loadTileLayers() {
      const layers = [];
      this.cBaseLayers.forEach(layer => {
        const newLayer = L.tileLayer(layer.url, layer.configParameters);
        layers.push(newLayer);
      });
      this.layersMap = layers;
    },
    setMap(layersToSetMap) {
      const bounds = this.setMapBounds();
      this.myMap = L.map("mapid", {
        center: this.cMapDefault.center,
        zoom: this.cMapDefault.zoom,
        minZoom: this.cMapDefault.minZoom,
        layers: layersToSetMap,
        maxBounds: bounds,
        maxBoundsViscosity: 1.0,
        worldCopyJump: true
      }).setView([this.cCenter.lat, this.cCenter.lon], this.cCenter.zoom);
    },
    setMapBounds() {
      return L.latLngBounds(L.latLng(-90, -Infinity), L.latLng(90, Infinity));
    },
    addGlobalSearchControl() {
      const self = this;
      L.easyButton({
        states: [
          {
            stateName: "search",
            icon: "fa-globe search",
            title: "Click to search without limits by coords",
            onClick: function() {
              self.myMap.fire("search");
            }
          }
        ]
      }).addTo(this.myMap);
    },
    addDrawControls() {
      this.customTextsDraw();
      this.editableLayers = new L.FeatureGroup();
      this.cDrawOptions.edit.featureGroup = this.editableLayers;
      this.myMap.addControl(new L.Control.Draw(this.cDrawOptions));
      this.addEditableLayerToMap();
    },
    addEditableLayerToMap() {
      this.myMap.addLayer(this.editableLayers);
    },
    customTextsDraw() {
      L.drawLocal.draw.toolbar.buttons.rectangle =
        "Draw a rectangle on the map to search inside thats coords ";
      L.drawLocal.draw.toolbar.buttons.marker =
        "Set a position marker on the map to search in thats coords";
      L.drawLocal.draw.handlers.marker.tooltip.start =
        "Click to search data in this coords";
      L.drawLocal.draw.handlers.simpleshape.tooltip.end =
        "Release mouse to search data inside the rectangle";
    },
    setupSwitcherLayer() {
      if (this.cSwitchLayerBaseMaps.length > 0) {
        const baseLayers = {};
        this.cSwitchLayerBaseMaps.forEach(baseLayer => {
          const newLayer = this.layersMap.find(
            layer => layer.options.id == baseLayer.idLayer
          );
          baseLayers[baseLayer.displayLayerName] = newLayer;
        });
        L.control.layers(baseLayers).addTo(this.myMap);
      }
    },
    setMarker() {
      this.layerMarkers = L.featureGroup().addTo(this.myMap);
      if (this.cMarkers.length > 0) {
        this.cMarkers.forEach(cMarker => {
          this.createLayerMarker(cMarker);
        });
      }
    },
    updateLayerMarker() {
      if (this.cMarkers.length > 0) this.addNewLayerMarkers();
      if (this.layerMarkers.length != 0) this.removeOldLayerMarkers();
    },
    addNewLayerMarkers() {
      this.cMarkers.forEach(cMarker => {
        this.checkMarkerExist(cMarker.mapID)
          ? ""
          : this.createLayerMarker(cMarker);
      });
    },
    moveAndZoomToMarkerPosition(mark) {
      mark.type == "extent"
        ? this.myMap.fitBounds(mark.getBounds())
        : this.myMap.fitBounds(panZoomMarker(mark), { maxZoom: 11 });
    },
    moveToMarkerPosition(mark) {
      mark.type == "extent"
        ? this.myMap.panTo(mark.getBounds().getCenter())
        : this.myMap.panTo(mark.getLatLng());
    },
    checkMarkerExist(id) {
      let exist = false;
      this.layerMarkers.eachLayer(layer => {
        if (layer.layerID === id) exist = true;
      });
      return exist;
    },
    createLayerMarker(cMarker) {
      let mark = {};
      switch (cMarker.type) {
        case "extent":
          mark = L.geoJSON(cMarker.extent, {
            style: { opacity: 1, color: cMarker.color }
          });
          mark.layerID = cMarker.mapID;
          mark.type = cMarker.type;
          break;
        case "marker":
          mark = L.marker(
            { lon: cMarker.longitude, lat: cMarker.latitude },
            { icon: createCustomIcon(cMarker.typeSensor, cMarker.color) }
          );
          mark.layerID = cMarker.mapID;
          mark.type = cMarker.type;
          break;
      }
      mark.addTo(this.layerMarkers);
      this.moveToMarkerPosition(mark);
    },
    removeOldLayerMarkers() {
      const self = this;
      this.layerMarkers.eachLayer(layer => {
        this.checkLayerMarkerDontExist(layer.layerID)
          ? self.layerMarkers.removeLayer(layer)
          : "";
      });
    },
    checkLayerMarkerDontExist(id) {
      return this.cMarkers.find(cMarker => cMarker.mapID === id) ? false : true;
    },
    setupListeners() {
      this.ListenerSearch();
      this.listenerDraw();
    },
    ListenerSearch() {
      const self = this;
      this.myMap.on("search", function() {
        self.removeOldEditableLayers();
        self.$emit("search-global");
      });
    },
    listenerDraw() {
      const self = this;
      this.myMap.on("draw:created", function(e) {
        const type = e.layerType;
        const markerLayer = e.layer;
        self.addDrawLayer(markerLayer, type);
        self.launchEventCoords(markerLayer, type);
      });
    },
    addDrawLayer(markerLayer, type) {
      this.removeOldEditableLayers();
      const styledLayer = this.customStyleLayer(markerLayer, type);
      this.editableLayers.addLayer(styledLayer);
    },
    removeOldEditableLayers() {
      this.editableLayers.eachLayer(layer => {
        this.editableLayers.removeLayer(layer);
      });
    },
    customStyleLayer(markerLayer, type) {
      if (type === "marker") {
        return markerLayer;
      }
      if (type === "rectangle") {
        return markerLayer.setStyle({
          weight: 5,
          fillOpacity: 0,
          dashArray: 10
        });
      }
    },
    launchEventCoords(markerLayer, type) {
      if (type === "marker") {
        const geoJson = markerLayer.toGeoJSON();
        const lon = geoJson.geometry.coordinates[0].toFixed(6);
        const lat = geoJson.geometry.coordinates[1].toFixed(6);
        this.$emit("marker-drawn", { lon: lon, lat: lat });
      }
      if (type === "rectangle") {
        const geoJson = markerLayer.toGeoJSON();
        const cords = geoJson.geometry.coordinates[0];
        const lonMin = cords[0][0].toFixed(6);
        const lonMax = cords[2][0].toFixed(6);
        const latMin = cords[3][1].toFixed(6);
        const latMax = cords[1][1].toFixed(6);
        this.$emit("rectangle-drawn", {
          lonMin: lonMin,
          lonMax: lonMax,
          latMin: latMin,
          latMax: latMax
        });
      }
    },
    goTo(marker) {
      let tempMark;
      if (marker.type === "extent") {
        tempMark = L.geoJSON(marker.coords);
        tempMark.type = marker.type;
      }
      if (marker.type === "marker") {
        tempMark = L.marker({ lon: marker.coords.lon, lat: marker.coords.lat });
        tempMark.type = marker.type;
      }
      this.moveAndZoomToMarkerPosition(tempMark);
    }
  }
};
</script>
<style src="leaflet/dist/leaflet.css"></style>
<style src="leaflet-draw/dist/leaflet.draw.css"></style>
<style
  src="leaflet-extra-markers/dist/css/leaflet.extra-markers.min.css"
></style>
<style scoped lang="scss">
.l-map {
  height: 100%;

  &__view {
    height: 100%;
    z-index: 0;
  }
}
</style>
