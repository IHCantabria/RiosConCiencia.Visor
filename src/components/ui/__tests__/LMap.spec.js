import { shallowMount } from "@vue/test-utils";
import LMap from "../LMap.vue";
import merge from "lodash.merge";

function createWrapper(overrides) {
  const defaultMountingOptions = {
    propsData: {
      cCenter: {
        lat: 0,
        lon: 0,
        zoom: 2,
        minZoom: 2
      },
      cBaseLayers: [
        {
          name: "carto-positron",
          url:
            "https://cartodb-basemaps-{s}.global.ssl.fastly.net/light_all/{z}/{x}/{y}{r}.png",
          configParameters: {
            id: "cartoPositronID",
            attribution:
              '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> &copy; <a href="https://cartodb.com/attributions">CartoDB</a>',
            subdomains: "abcd",
            minZoom: 2,
            maxZoom: 19
          }
        },
        {
          name: "carto-dark",
          url:
            "https://cartodb-basemaps-{s}.global.ssl.fastly.net/dark_all/{z}/{x}/{y}{r}.png",
          configParameters: {
            id: "cartoDarkID",
            attribution:
              '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> &copy; <a href="https://cartodb.com/attributions">CartoDB</a>',
            subdomains: "abcd",
            minZoom: 2,
            maxZoom: 19
          }
        },
        {
          name: "carto-voyager",
          url:
            "https://cartodb-basemaps-{s}.global.ssl.fastly.net/rastertiles/voyager_labels_under/{z}/{x}/{y}{r}.png",
          configParameters: {
            id: "cartoVoyagerID",
            attribution:
              '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> &copy; <a href="https://cartodb.com/attributions">CartoDB</a>',
            subdomains: "abcd",
            minZoom: 2,
            maxZoom: 19
          }
        },
        {
          name: "statemen-watercolor",
          url:
            "https://stamen-tiles-{s}.a.ssl.fastly.net/watercolor/{z}/{x}/{y}.{ext}",
          configParameters: {
            id: "statemenWatercolorID",
            attribution:
              'Map tiles by <a href="https://stamen.com">Stamen Design</a>, <a href="https://creativecommons.org/licenses/by/3.0">CC BY 3.0</a> &mdash; Map data &copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a>',
            subdomains: "abcd",
            minZoom: 2,
            maxZoom: 16,
            ext: "png"
          }
        },
        {
          name: "osm-mapnik",
          url: "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
          configParameters: {
            id: "osmMapnikID",
            attribution:
              '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a>',
            minZoom: 2,
            maxZoom: 19
          }
        },
        {
          name: "esri-imagery",
          url:
            "https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}",
          configParameters: {
            id: "esriImageryID",
            attribution:
              "Tiles &copy; Esri &mdash; Source: Esri, i-cubed, USDA, USGS, AEX, GeoEye, Getmapping, Aerogrid, IGN, IGP, UPR-EGP, and the GIS User Community",
            minZoom: 2
          }
        },
        {
          name: "esri-oceans",
          url:
            "https://server.arcgisonline.com/ArcGIS/rest/services/Ocean_Basemap/MapServer/tile/{z}/{y}/{x}",
          configParameters: {
            id: "esriOceansID",
            attribution:
              "Tiles &copy; Esri &mdash; Sources: GEBCO, NOAA, CHS, OSU, UNH, CSUMB, National Geographic, DeLorme, NAVTEQ, and Esri",
            minZoom: 2,
            maxZoom: 13
          }
        },
        {
          name: "esri-gray",
          url:
            "https://server.arcgisonline.com/ArcGIS/rest/services/Canvas/World_Light_Gray_Base/MapServer/tile/{z}/{y}/{x}",
          configParameters: {
            id: "esriGrayID",
            attribution: "Tiles &copy; Esri &mdash; Esri, DeLorme, NAVTEQ",
            minZoom: 2,
            maxZoom: 16
          }
        }
      ],
      cMapDefault: {
        center: [0, 0],
        zoom: 2,
        layers: ["cartoPositronID"]
      },
      cSwitchLayerBaseMaps: [
        {
          displayLayerName: "carto-positron",
          idLayer: "cartoPositronID"
        },
        {
          displayLayerName: "carto-dark",
          idLayer: "cartoDarkID"
        },
        {
          displayLayerName: "carto-voyager",
          idLayer: "cartoVoyagerID"
        },
        {
          displayLayerName: "statemen-watercolor",
          idLayer: "statemenWatercolorID"
        },
        {
          displayLayerName: "osm-mapnik",
          idLayer: "osmMapnikID"
        },
        {
          displayLayerName: "esri-imagery",
          idLayer: "esriImageryID"
        },
        {
          displayLayerName: "esri-oceans",
          idLayer: "esriOceansID"
        },
        {
          displayLayerName: "esri-gray",
          idLayer: "esriGrayID"
        }
      ],
      cDrawOptions: {
        draw: {
          polyline: false,
          polygon: false,
          rectangle: {},
          circle: false,
          circlemarker: false,
          marker: {}
        },
        edit: {
          edit: false,
          options: {},
          remove: false
        },
        position: "topleft"
      },
      cMarkers: []
    }
  };
  return shallowMount(LMap, merge(defaultMountingOptions, overrides));
}

describe("lMap.vue", () => {
  var mapWrapper;
  function mapDivCreate() {
    mapWrapper = createWrapper({ attachToDocument: true });
  }
  function mapDivDelete() {
    mapWrapper.destroy();
  }
  beforeEach(() => {
    mapDivCreate();
  });
  afterEach(() => {
    mapDivDelete();
  });
  //UNIT TESTS
  it("renders a map", () => {
    expect.assertions(1);
    const divMap = mapWrapper.find(`#mapid`);
    expect(divMap.exists()).toBe(true);
  });
  it("define baseLayers onload with loadTileLayers", () => {
    expect.assertions(1);
    const layersMap = mapWrapper.vm.layersMap;
    expect(layersMap).not.toBe([]);
  });
  it("define map onload with setMap", () => {
    expect.assertions(1);
    const myMap = mapWrapper.vm.myMap;
    expect(myMap).toBeDefined();
  });
  it("add search Control onload with addGlobalSearchControl", () => {
    expect.assertions(1);
    const searchControl = mapWrapper.find(".easy-button-container");
    expect(searchControl.exists()).toBe(true);
  });
  it("add Draw Control onload with addDrawControls", () => {
    expect.assertions(1);
    const drawControl = mapWrapper.find(`.leaflet-draw`);
    expect(drawControl.exists()).toBe(true);
  });
  it("add SwitcherLayer Control onload with setupSwitcherLayer", () => {
    expect.assertions(1);
    const switcherControl = mapWrapper.find(`.leaflet-control-layers`);
    expect(switcherControl.exists()).toBe(true);
  });
  it("emit event after draw a rectangle on map", () => {
    expect.assertions(1);
    const map = mapWrapper.vm.myMap;
    const bounds = [
      [54.559322, -5.767822],
      [56.1210604, -3.02124]
    ];
    const layerTest = L.rectangle(bounds, {
      color: "#ff7800",
      weight: 1
    });
    jest.spyOn(mapWrapper.vm, "addDrawLayer").mockImplementation();
    map.fire("draw:created", { layerType: "rectangle", layer: layerTest });
    expect(mapWrapper.emitted("rectangle-drawn")).toStrictEqual([
      [
        {
          lonMin: "-5.767822",
          lonMax: "-3.021240",
          latMin: "54.559322",
          latMax: "56.121060"
        }
      ]
    ]);
  });
  it("call launchEventCoords and addDrawLayer after draw a rectangle on map", () => {
    expect.assertions(2);
    const map = mapWrapper.vm.myMap;
    const bounds = [
      [54.559322, -5.767822],
      [56.1210604, -3.02124]
    ];
    const layerTest = L.rectangle(bounds, {
      color: "#ff7800",
      weight: 1
    });
    const mockAddDrawLayer = jest.fn();
    mapWrapper.vm.addDrawLayer = mockAddDrawLayer;
    const spyLaunchEventCoords = jest.spyOn(mapWrapper.vm, "launchEventCoords");
    map.fire("draw:created", { layerType: "rectangle", layer: layerTest });
    expect(mockAddDrawLayer).toHaveBeenCalled();
    expect(spyLaunchEventCoords).toHaveBeenCalled();
  });
  it("emit event after place a marker on map", () => {
    expect.assertions(1);
    const map = mapWrapper.vm.myMap;
    const bounds = [54.559322, -5.767822];
    const layerTest = L.marker(bounds, {
      color: "#ff7800",
      weight: 1
    });
    map.fire("draw:created", { layerType: "marker", layer: layerTest });
    expect(mapWrapper.emitted("marker-drawn")).toStrictEqual([
      [
        {
          lat: "54.559322",
          lon: "-5.767822"
        }
      ]
    ]);
  });
  it("call launchEventCoords after place a marker on map", () => {
    expect.assertions(1);
    const map = mapWrapper.vm.myMap;
    const bounds = [54.559322, -5.767822];
    const layerTest = L.marker(bounds, {
      color: "#ff7800",
      weight: 1
    });
    const spyLaunchEventCoords = jest.spyOn(mapWrapper.vm, "launchEventCoords");
    map.fire("draw:created", { layerType: "marker", layer: layerTest });
    expect(spyLaunchEventCoords).toHaveBeenCalled();
  });
  it("emit event after click search button", () => {
    expect.assertions(1);
    const map = mapWrapper.vm.myMap;
    map.fire("search");
    expect(mapWrapper.emitted("search-global")).toStrictEqual([[]]);
  });
  it("init method called after mounted", async () => {
    expect.assertions(1);
    const mockInit = jest.fn();
    const wrapper = createWrapper({
      methods: {
        initMap: mockInit
      }
    });
    await wrapper.vm.$nextTick();
    expect(mockInit).toHaveBeenCalled();
  });
  //SNAPSHOT TESTS
  it("renders correctly", () => {
    expect.assertions(1);
    expect(mapWrapper.element).toMatchSnapshot();
  });
});
