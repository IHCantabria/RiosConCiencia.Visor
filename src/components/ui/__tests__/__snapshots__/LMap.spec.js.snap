// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lMap.vue renders correctly 1`] = `
<div
  class="l-map"
>
  <div
    class="l-map__view leaflet-container leaflet-grab leaflet-touch-drag"
    id="mapid"
    style="position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      >
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 19; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://cartodb-basemaps-c.global.ssl.fastly.net/light_all/2/1/1.png"
              style="width: 256px; height: 256px; left: -256px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://cartodb-basemaps-d.global.ssl.fastly.net/light_all/2/2/1.png"
              style="width: 256px; height: 256px; left: 0px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://cartodb-basemaps-d.global.ssl.fastly.net/light_all/2/1/2.png"
              style="width: 256px; height: 256px; left: -256px; top: 0px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://cartodb-basemaps-a.global.ssl.fastly.net/light_all/2/2/2.png"
              style="width: 256px; height: 256px; left: 0px; top: 0px;"
            />
          </div>
        </div>
      </div>
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            +
          </a>
          <a
            aria-label="Zoom out"
            class="leaflet-control-zoom-out leaflet-disabled"
            href="#"
            role="button"
            title="Zoom out"
          >
            −
          </a>
        </div>
        <div
          class="leaflet-bar easy-button-container leaflet-control"
        >
          <button
            class="easy-button-button leaflet-bar-part leaflet-interactive search-active"
            title="Click to search without limits by coords"
            type="button"
          >
            <span
              class="button-state state-search search-active"
            >
              <span
                class="fa fa-globe search"
              />
            </span>
          </button>
        </div>
        <div
          class="leaflet-draw leaflet-control"
        >
          <div
            class="leaflet-draw-section"
          >
            <div
              class="leaflet-draw-toolbar leaflet-bar leaflet-draw-toolbar-top"
            >
              <a
                class="leaflet-draw-draw-rectangle"
                href="#"
                title="Draw a rectangle on the map to search inside thats coords "
              >
                <span
                  class="sr-only"
                >
                  Draw a rectangle on the map to search inside thats coords 
                </span>
              </a>
              <a
                class="leaflet-draw-draw-marker"
                href="#"
                title="Set a position marker on the map to search in thats coords"
              >
                <span
                  class="sr-only"
                >
                  Set a position marker on the map to search in thats coords
                </span>
              </a>
            </div>
            <ul
              class="leaflet-draw-actions"
            />
          </div>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      >
        <div
          aria-haspopup="true"
          class="leaflet-control-layers leaflet-control"
        >
          <a
            class="leaflet-control-layers-toggle"
            href="#"
            title="Layers"
          />
          <section
            class="leaflet-control-layers-list"
          >
            <div
              class="leaflet-control-layers-base"
            >
              <label>
                <div>
                  <input
                    checked="checked"
                    class="leaflet-control-layers-selector"
                    name="leaflet-base-layers_436"
                    type="radio"
                  />
                  <span>
                     carto-positron
                  </span>
                </div>
              </label>
              <label>
                <div>
                  <input
                    class="leaflet-control-layers-selector"
                    name="leaflet-base-layers_436"
                    type="radio"
                  />
                  <span>
                     carto-dark
                  </span>
                </div>
              </label>
              <label>
                <div>
                  <input
                    class="leaflet-control-layers-selector"
                    name="leaflet-base-layers_436"
                    type="radio"
                  />
                  <span>
                     carto-voyager
                  </span>
                </div>
              </label>
              <label>
                <div>
                  <input
                    class="leaflet-control-layers-selector"
                    name="leaflet-base-layers_436"
                    type="radio"
                  />
                  <span>
                     statemen-watercolor
                  </span>
                </div>
              </label>
              <label>
                <div>
                  <input
                    class="leaflet-control-layers-selector"
                    name="leaflet-base-layers_436"
                    type="radio"
                  />
                  <span>
                     osm-mapnik
                  </span>
                </div>
              </label>
              <label>
                <div>
                  <input
                    class="leaflet-control-layers-selector"
                    name="leaflet-base-layers_436"
                    type="radio"
                  />
                  <span>
                     esri-imagery
                  </span>
                </div>
              </label>
              <label>
                <div>
                  <input
                    class="leaflet-control-layers-selector"
                    name="leaflet-base-layers_436"
                    type="radio"
                  />
                  <span>
                     esri-oceans
                  </span>
                </div>
              </label>
              <label>
                <div>
                  <input
                    class="leaflet-control-layers-selector"
                    name="leaflet-base-layers_436"
                    type="radio"
                  />
                  <span>
                     esri-gray
                  </span>
                </div>
              </label>
            </div>
            <div
              class="leaflet-control-layers-separator"
              style="display: none;"
            />
            <div
              class="leaflet-control-layers-overlays"
            />
          </section>
        </div>
      </div>
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JS library for interactive maps"
          >
            Leaflet
          </a>
           | © 
          <a
            href="https://www.openstreetmap.org/copyright"
          >
            OpenStreetMap
          </a>
           © 
          <a
            href="https://cartodb.com/attributions"
          >
            CartoDB
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
